File: crates/fc-scheduler/src/poller.rs
//! Pending job poller
//!
//! Polls for PENDING dispatch jobs, groups by message_group, filters by
//! dispatch mode and connection status, then submits to the MessageGroupDispatcher
//! for ordered, semaphore-limited dispatch.

use std::collections::{HashMap, HashSet};
use std::sync::Arc;
use std::time::{Duration, Instant};

use parking_lot::RwLock;
use sea_orm::{
    DatabaseBackend, DatabaseConnection, FromQueryResult, Statement,
};
use tracing::{debug, info, trace, warn};

use crate::{
    BlockOnErrorChecker, DispatchJob, DispatchMode,
    MessageGroupDispatcher, SchedulerConfig, SchedulerError,
};

const DEFAULT_MESSAGE_GROUP: &str = "default";

/// Caches the set of subscription IDs whose connections are PAUSED.
/// Refreshed in the background on a configurable interval (default 60s).
/// Avoids a JOIN query on every 5-second poll cycle.
#[derive(Clone)]
pub struct PausedConnectionCache {
    db: DatabaseConnection,
    /// Cached set of paused subscription IDs
    cache: Arc<RwLock<HashSet<String>>>,
    /// When the cache was last refreshed
    last_refresh: Arc<RwLock<Instant>>,
    /// How long before the cache is considered stale
    ttl: Duration,
}

impl PausedConnectionCache {
    pub fn new(db: DatabaseConnection, ttl: Duration) -> Self {
        Self {
            db,
            cache: Arc::new(RwLock::new(HashSet::new())),
            last_refresh: Arc::new(RwLock::new(Instant::now() - ttl - Duration::from_secs(1))), // force initial refresh
            ttl,
        }
    }

    /// Get the cached set of paused subscription IDs.
    /// If the cache is stale, refreshes synchronously before returning.
    pub async fn get_paused_subscription_ids(&self) -> Result<HashSet<String>, SchedulerError> {
        if self.last_refresh.read().elapsed() > self.ttl {
            self.refresh().await?;
        }
        Ok(self.cache.read().clone())
    }

    /// Refresh the cache from the database.
    pub async fn refresh(&self) -> Result<(), SchedulerError> {
        let sql = "SELECT s.id FROM msg_subscriptions s \
                   JOIN msg_connections c ON c.id = s.connection_id \
                   WHERE c.status = 'PAUSED'";

        #[derive(FromQueryResult)]
        struct SubIdRow { id: String }

        let rows = SubIdRow::find_by_statement(
            Statement::from_sql_and_values(DatabaseBackend::Postgres, sql, vec![]),
        )
        .all(&self.db)
        .await?;

        let paused: HashSet<String> = rows.into_iter().map(|r| r.id).collect();
        let count = paused.len();

        *self.cache.write() = paused;
        *self.last_refresh.write() = Instant::now();

        debug!(paused_subscriptions = count, "Refreshed paused connection cache");
        Ok(())
    }

    /// Spawn a background task that refreshes the cache periodically.
    pub fn spawn_refresh_task(&self, shutdown: tokio::sync::broadcast::Sender<()>) {
        let cache = self.clone();
        let interval = self.ttl;
        let mut shutdown_rx = shutdown.subscribe();

        tokio::spawn(async move {
            let mut ticker = tokio::time::interval(interval);
            // Skip first tick (handled by initial get_paused_subscription_ids call)
            ticker.tick().await;

            loop {
                tokio::select! {
                    _ = ticker.tick() => {
                        if let Err(e) = cache.refresh().await {
                            warn!(error = %e, "Failed to refresh paused connection cache");
                        }
                    }
                    _ = shutdown_rx.recv() => {
                        info!("Paused connection cache refresh task shutting down");
                        break;
                    }
                }
            }
        });
    }
}

#[derive(Clone)]
pub struct PendingJobPoller {
    config: SchedulerConfig,
    db: DatabaseConnection,
    block_checker: Arc<BlockOnErrorChecker>,
    group_dispatcher: Arc<MessageGroupDispatcher>,
    paused_cache: Arc<PausedConnectionCache>,
}

impl PendingJobPoller {
    pub fn new(
        config: SchedulerConfig,
        db: DatabaseConnection,
        group_dispatcher: Arc<MessageGroupDispatcher>,
    ) -> Self {
        let block_checker = Arc::new(BlockOnErrorChecker::new(db.clone()));
        let paused_cache = Arc::new(PausedConnectionCache::new(
            db.clone(),
            Duration::from_secs(60), // 60s TTL
        ));
        Self {
            config,
            db,
            block_checker,
            group_dispatcher,
            paused_cache,
        }
    }

    /// Get the paused connection cache (for spawning the background refresh task)
    pub fn paused_cache(&self) -> &Arc<PausedConnectionCache> {
        &self.paused_cache
    }

    /// Poll for pending jobs and submit them for dispatch.
    /// Returns the number of jobs found (used for adaptive polling).
    pub async fn poll(&self) -> Result<usize, SchedulerError> {
        let pending_jobs = self.find_pending_jobs().await?;
        if pending_jobs.is_empty() {
            trace!("No pending jobs found");
            return Ok(0);
        }

        let job_count = pending_jobs.len();
        debug!(count = job_count, "Found pending jobs to process");
        metrics::gauge!("scheduler.pending_jobs").set(job_count as f64);

        // Group by message_group
        let jobs_by_group = Self::group_by_message_group(pending_jobs);
        let groups: HashSet<String> = jobs_by_group.keys().cloned().collect();

        // Batch check for blocked groups (single query)
        let blocked_groups = self.block_checker.get_blocked_groups(&groups).await?;
        metrics::gauge!("scheduler.blocked_groups").set(blocked_groups.len() as f64);

        // Process each group
        for (group, jobs) in jobs_by_group {
            if blocked_groups.contains(&group) {
                debug!(group = %group, count = jobs.len(), "Message group blocked, skipping");
                metrics::counter!("scheduler.jobs.blocked_total").increment(jobs.len() as u64);
                continue;
            }

            // Filter by dispatch mode
            let dispatchable = Self::filter_by_dispatch_mode(jobs, &blocked_groups);
            if !dispatchable.is_empty() {
                debug!(group = %group, count = dispatchable.len(), "Submitting jobs for message group");
                // Submit to the group dispatcher (1-in-flight per group, semaphore-limited)
                self.group_dispatcher.submit_jobs(&group, dispatchable);
            }
        }
        Ok(job_count)
    }

    /// Query PENDING jobs with proper ordering: message_group, sequence, created_at.
    /// Connection filtering is applied post-query (matching TS ConnectionCache pattern).
    async fn find_pending_jobs(&self) -> Result<Vec<DispatchJob>, SchedulerError> {
        let sql = "SELECT id, message_group, dispatch_pool_id, status, mode, target_url, \
                    payload, sequence, created_at, updated_at, queued_at, last_error, subscription_id \
             FROM msg_dispatch_jobs \
             WHERE status = 'PENDING' \
             ORDER BY message_group ASC NULLS LAST, sequence ASC, created_at ASC \
             LIMIT $1";

        let jobs = DispatchJob::find_by_statement(
            Statement::from_sql_and_values(
                DatabaseBackend::Postgres,
                sql,
                vec![sea_orm::Value::from(self.config.batch_size as i64)],
            ),
        )
        .all(&self.db)
        .await?;

        if !self.config.connection_filter_enabled || jobs.is_empty() {
            return Ok(jobs);
        }

        // Filter using cached paused connection set (refreshed every 60s)
        let paused_sub_ids = self.paused_cache.get_paused_subscription_ids().await?;
        if paused_sub_ids.is_empty() {
            return Ok(jobs);
        }

        let filtered: Vec<DispatchJob> = jobs.into_iter()
            .filter(|j| {
                match &j.subscription_id {
                    Some(sid) => !paused_sub_ids.contains(sid),
                    None => true,
                }
            })
            .collect();

        Ok(filtered)
    }

    fn group_by_message_group(jobs: Vec<DispatchJob>) -> HashMap<String, Vec<DispatchJob>> {
        let mut grouped: HashMap<String, Vec<DispatchJob>> = HashMap::new();
        for job in jobs {
            let group = job.message_group.clone().unwrap_or_else(|| DEFAULT_MESSAGE_GROUP.to_string());
            grouped.entry(group).or_default().push(job);
        }
        grouped
    }

    fn filter_by_dispatch_mode(jobs: Vec<DispatchJob>, blocked_groups: &HashSet<String>) -> Vec<DispatchJob> {
        jobs.into_iter()
            .filter(|job| {
                let group = job.message_group.as_deref().unwrap_or(DEFAULT_MESSAGE_GROUP);
                match job.dispatch_mode() {
                    DispatchMode::Immediate => true,
                    DispatchMode::NextOnError | DispatchMode::BlockOnError => {
                        !blocked_groups.contains(group)
                    }
                }
            })
            .collect()
    }
}

File: internal/platform/scheduler/poller.go
package scheduler

import (
	"context"
	"log/slog"
	"sync"
	"time"

	"github.com/jackc/pgx/v5"
	"github.com/jackc/pgx/v5/pgxpool"

	"github.com/flowcatalyst/flowcatalyst-go/internal/common"
)

// defaultMessageGroup is the grouping key for jobs without a
// message_group — mirrors Rust's DEFAULT_MESSAGE_GROUP (poller.rs).
const defaultMessageGroup = "default"

// PausedConnectionCache caches the set of subscription IDs whose target
// connections are PAUSED. The poller filters jobs whose subscription
// matches; those jobs sit in PENDING until the connection is reactivated.
type PausedConnectionCache struct {
	pool *pgxpool.Pool
	ttl  time.Duration

	mu          sync.RWMutex
	paused      map[string]struct{}
	lastRefresh time.Time
}

// NewPausedConnectionCache wires the cache.
func NewPausedConnectionCache(pool *pgxpool.Pool, ttl time.Duration) *PausedConnectionCache {
	return &PausedConnectionCache{
		pool:        pool,
		ttl:         ttl,
		paused:      make(map[string]struct{}),
		lastRefresh: time.Now().Add(-2 * ttl), // force initial refresh
	}
}

// PausedSubscriptionIDs returns the cached set, refreshing if stale.
func (c *PausedConnectionCache) PausedSubscriptionIDs(ctx context.Context) (map[string]struct{}, error) {
	c.mu.RLock()
	if time.Since(c.lastRefresh) < c.ttl {
		out := make(map[string]struct{}, len(c.paused))
		for k := range c.paused {
			out[k] = struct{}{}
		}
		c.mu.RUnlock()
		return out, nil
	}
	c.mu.RUnlock()
	if err := c.refresh(ctx); err != nil {
		return nil, err
	}
	c.mu.RLock()
	defer c.mu.RUnlock()
	out := make(map[string]struct{}, len(c.paused))
	for k := range c.paused {
		out[k] = struct{}{}
	}
	return out, nil
}

func (c *PausedConnectionCache) refresh(ctx context.Context) error {
	rows, err := c.pool.Query(ctx,
		`SELECT s.id FROM msg_subscriptions s
		   JOIN msg_connections c ON c.id = s.connection_id
		  WHERE c.status = 'PAUSED'`)
	if err != nil {
		return err
	}
	defer rows.Close()
	paused := make(map[string]struct{})
	for rows.Next() {
		var id string
		if err := rows.Scan(&id); err != nil {
			return err
		}
		paused[id] = struct{}{}
	}
	if err := rows.Err(); err != nil {
		return err
	}
	c.mu.Lock()
	c.paused = paused
	c.lastRefresh = time.Now()
	c.mu.Unlock()
	slog.Debug("paused connection cache refreshed", "paused_subscriptions", len(paused))
	return nil
}

// PendingJobPoller polls msg_dispatch_jobs for PENDING jobs ready to
// dispatch (next_retry_at <= NOW or null), filters them through the
// pause + block-on-error checks, and submits to the MessageGroupDispatcher.
type PendingJobPoller struct {
	cfg         Config
	pool        *pgxpool.Pool
	dispatcher  *MessageGroupDispatcher
	pausedCache *PausedConnectionCache
	// IsLeader gates claiming: when non-nil and false, the poller idles.
	// The per-group FIFO dispatcher is in-process only, so within-group
	// ordering requires a single active scheduler — concurrent SKIP-LOCKED
	// claims across replicas would dispatch a group's jobs out of order.
	// nil = always run (standby disabled). Set by Scheduler.Run.
	IsLeader func() bool
}

// NewPendingJobPoller wires the poller.
func NewPendingJobPoller(cfg Config, pool *pgxpool.Pool, dispatcher *MessageGroupDispatcher, pausedCache *PausedConnectionCache) *PendingJobPoller {
	return &PendingJobPoller{cfg: cfg, pool: pool, dispatcher: dispatcher, pausedCache: pausedCache}
}

// Run drives the poller until ctx is cancelled.
func (p *PendingJobPoller) Run(ctx context.Context) {
	tick := time.NewTicker(p.cfg.PollInterval)
	defer tick.Stop()
	slog.Info("dispatch job poller starting", "interval", p.cfg.PollInterval, "batch_size", p.cfg.BatchSize)
	for {
		select {
		case <-ctx.Done():
			slog.Info("dispatch job poller stopped")
			return
		case <-tick.C:
			if p.IsLeader != nil && !p.IsLeader() {
				continue // only the leader claims
			}
			if err := p.pollOnce(ctx); err != nil {
				slog.Warn("poll error", "err", err)
			}
		}
	}
}

// pollOnce claims a batch of jobs and submits them to the dispatcher.
func (p *PendingJobPoller) pollOnce(ctx context.Context) error {
	paused, err := p.pausedCache.PausedSubscriptionIDs(ctx)
	if err != nil {
		return err
	}
	tx, err := p.pool.Begin(ctx)
	if err != nil {
		return err
	}
	defer func() { _ = tx.Rollback(ctx) }()

	// Claim PENDING jobs ready for dispatch. SKIP LOCKED so multiple
	// scheduler instances don't contend.
	// Retry timing is owned by the dispatcher's backoff loop, not by a
	// scheduled-for column on the row — matches the Rust poller in
	// crates/fc-platform/src/scheduler/poller.rs. The embedded schema
	// has neither `next_retry_at` (only added in migration 011's
	// no-op-on-embedded CREATE TABLE IF NOT EXISTS) nor a scheduled-
	// filtered claim path.
	// scheduled_for gates retry backoff: /api/dispatch/process reschedules a
	// failed job to NOW()+backoff (status back to PENDING) and ACKs the queue
	// message, so the poller is the single re-dispatch driver — no queue-NACK
	// racing the poll. A NULL scheduled_for (every freshly-created job) is
	// always eligible.
	rows, err := tx.Query(ctx,
		`SELECT id, subscription_id, message_group, mode, attempt_count, target_url
		   FROM msg_dispatch_jobs
		  WHERE status = 'PENDING'
		    AND (scheduled_for IS NULL OR scheduled_for <= NOW())
		  ORDER BY message_group ASC NULLS LAST, sequence ASC, created_at ASC
		  LIMIT $1
		  FOR UPDATE SKIP LOCKED`,
		p.cfg.BatchSize)
	if err != nil {
		return err
	}
	var claims []dispatchClaim
	for rows.Next() {
		var c dispatchClaim
		var msgGroup *string
		var subID *string
		if err := rows.Scan(&c.id, &subID, &msgGroup, &c.mode, &c.attempt, &c.target); err != nil {
			rows.Close()
			return err
		}
		if subID != nil {
			c.subID = *subID
		}
		if msgGroup != nil {
			c.group = *msgGroup
		}
		claims = append(claims, c)
	}
	rows.Close()
	if len(claims) == 0 {
		return nil
	}

	// Filter and gather IDs to mark QUEUED. Dispatch is deliberately
	// deferred until the claim tx has committed: publishing while the tx was
	// still open meant (a) a commit failure after a publish re-claimed the
	// already-published job on the next poll (duplicate dispatch), and (b) a
	// publish failure's QUEUED→PENDING revert no-oped because the QUEUED
	// status it guards on hadn't committed yet (row stuck until stale
	// recovery).
	//
	// Filter order mirrors the Rust poll (poller.rs): paused-subscription
	// filter, then group, then the blocked-group hold-back, then the
	// per-mode filter. Skipped claims are simply left PENDING — their row
	// locks release at commit and the next poll retries them.
	live, skippedPaused := filterPausedSubscriptions(claims, paused)

	byGroup := groupByMessageGroup(live)
	candidates := make([]string, 0, len(byGroup))
	for g := range byGroup {
		candidates = append(candidates, g)
	}
	blocked, err := blockedGroups(ctx, tx, candidates)
	if err != nil {
		return err
	}

	var queued []string
	var tokens []DispatchJobToken
	skippedBlocked := 0
	for group, jobs := range byGroup {
		// A FAILED/ERROR sibling holds back the whole group this tick —
		// ordered jobs must not jump past the failure, and the operator
		// resolving it (retry/cancel) unblocks the group for the next
		// poll. Rust skips the group before its mode filter, so even
		// IMMEDIATE jobs in a blocked group wait; preserved 1:1.
		if _, isBlocked := blocked[group]; isBlocked {
			slog.Debug("message group blocked, skipping", "group", group, "count", len(jobs))
			skippedBlocked += len(jobs)
			continue
		}
		for _, c := range filterByDispatchMode(jobs, blocked) {
			queued = append(queued, c.id)
			tokens = append(tokens, DispatchJobToken{
				JobID:        c.id,
				MessageGroup: c.group,
				TargetURL:    c.target,
			})
		}
	}

	if len(queued) > 0 {
		if _, err := tx.Exec(ctx,
			`UPDATE msg_dispatch_jobs SET status = 'QUEUED', updated_at = NOW()
			  WHERE id = ANY($1)`, queued); err != nil {
			return err
		}
	}
	if err := tx.Commit(ctx); err != nil {
		return err
	}

	// QUEUED is durable — now hand the jobs to the dispatcher. A publish
	// failure reverts QUEUED→PENDING (dispatcher.dispatch), which is
	// guaranteed to see the committed status; a crash between commit and
	// Submit leaves rows QUEUED for stale recovery — the same failure mode
	// as a crash mid-publish, handled by the existing recovery loop.
	for _, t := range tokens {
		p.dispatcher.Submit(ctx, t)
	}

	if len(queued) > 0 || skippedPaused > 0 || skippedBlocked > 0 {
		slog.Debug("poll tick",
			"queued", len(queued),
			"skipped_paused", skippedPaused,
			"skipped_blocked", skippedBlocked)
	}
	return nil
}

// dispatchClaim is one PENDING row claimed by the poll query. group and
// subID are "" when the column is NULL.
type dispatchClaim struct {
	id, subID, group, mode, target string
	attempt                        int32
}

// messageGroupKey maps a claim's message_group to its grouping key: jobs
// without a group bucket under "default", mirroring Rust's
// group_by_message_group (poller.rs).
func messageGroupKey(group string) string {
	if group == "" {
		return defaultMessageGroup
	}
	return group
}

// groupByMessageGroup buckets claims by grouping key. The poll query's
// (message_group, sequence, created_at) order is preserved within each
// group — that order is what the dispatcher's per-group FIFO relies on.
func groupByMessageGroup(claims []dispatchClaim) map[string][]dispatchClaim {
	grouped := make(map[string][]dispatchClaim)
	for _, c := range claims {
		key := messageGroupKey(c.group)
		grouped[key] = append(grouped[key], c)
	}
	return grouped
}

// filterPausedSubscriptions drops claims whose subscription's connection
// is PAUSED; they sit in PENDING until the connection is reactivated.
// Claims without a subscription always pass. Returns the survivors and
// the dropped count.
func filterPausedSubscriptions(claims []dispatchClaim, paused map[string]struct{}) ([]dispatchClaim, int) {
	if len(paused) == 0 {
		return claims, 0
	}
	kept := make([]dispatchClaim, 0, len(claims))
	for _, c := range claims {
		if c.subID != "" {
			if _, isPaused := paused[c.subID]; isPaused {
				continue
			}
		}
		kept = append(kept, c)
	}
	return kept, len(claims) - len(kept)
}

// filterByDispatchMode keeps the claims whose mode allows dispatch given
// the blocked groups: IMMEDIATE always dispatches; NEXT_ON_ERROR and
// BLOCK_ON_ERROR hold back while their group is blocked. 1:1 port of
// Rust's filter_by_dispatch_mode (poller.rs) — including the lenient
// parse where unknown modes count as IMMEDIATE. The group-level skip in
// pollOnce makes this currently redundant (a blocked group never reaches
// it), but it's kept for fidelity and so a future relaxation of the
// group skip doesn't silently lose the per-mode semantics.
func filterByDispatchMode(claims []dispatchClaim, blocked map[string]struct{}) []dispatchClaim {
	kept := make([]dispatchClaim, 0, len(claims))
	for _, c := range claims {
		switch common.ParseDispatchMode(c.mode) {
		case common.DispatchNextOnError, common.DispatchBlockOnError:
			if _, isBlocked := blocked[messageGroupKey(c.group)]; isBlocked {
				continue
			}
		}
		kept = append(kept, c)
	}
	return kept
}

// blockedGroups returns the subset of candidate groups that currently
// hold a FAILED or ERROR job — one batch query per poll, the port of
// Rust's BlockOnErrorChecker (mod.rs). A NULL message_group can never
// block: `= ANY` never matches NULL, so a failed ungrouped job does not
// hold back the "default" bucket. Preserve that exactly — only a row
// whose message_group is literally 'default' blocks ungrouped jobs.
func blockedGroups(ctx context.Context, tx pgx.Tx, groups []string) (map[string]struct{}, error) {
	blocked := make(map[string]struct{})
	if len(groups) == 0 {
		return blocked, nil
	}
	rows, err := tx.Query(ctx,
		`SELECT DISTINCT message_group FROM msg_dispatch_jobs
		  WHERE message_group = ANY($1) AND status IN ('FAILED', 'ERROR')`,
		groups)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	for rows.Next() {
		var g string
		if err := rows.Scan(&g); err != nil {
			return nil, err
		}
		blocked[g] = struct{}{}
	}
	if err := rows.Err(); err != nil {
		return nil, err
	}
	return blocked, nil
}

// DispatchJobToken is the value the poller hands the dispatcher. It
// carries just enough to publish to the queue without re-reading the
// job row.
type DispatchJobToken struct {
	JobID        string
	MessageGroup string
	TargetURL    string
}
